core/src/commonMain/kotlin/com/powersync/attachments/AttachmentQueue.kt

```diff
@@ -0,0 +1,489 @@
+package com.powersync.attachments
+
+import co.touchlab.kermit.Logger
+import com.powersync.PowerSyncDatabase
+import com.powersync.PowerSyncException
+import com.powersync.attachments.implementation.AttachmentServiceImpl
+import com.powersync.attachments.storage.IOLocalStorageAdapter
+import com.powersync.attachments.sync.SyncingService
+import com.powersync.db.internal.ConnectionContext
+import com.powersync.db.runWrappedSuspending
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.IO
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.io.files.Path
+import kotlin.coroutines.cancellation.CancellationException
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+
+/**
+ * A watched attachment record item.
+ * This is usually returned from watching all relevant attachment IDs.
+ */
+public data class WatchedAttachmentItem(
+    /**
+     * Id for the attachment record
+     */
+    public val id: String,
+    /**
+     * File extension used to determine an internal filename for storage if no [filename] is provided
+     */
+    public val fileExtension: String? = null,
+    /**
+     * Filename to store the attachment with
+     */
+    public val filename: String? = null,
+    /**
+     * Optional meta data for the attachment record
+     */
+    public val metaData: String? = null,
+) {
+    init {
+        require(fileExtension != null || filename != null) {
+            "Either fileExtension or filename must be provided."
+        }
+    }
+}
+
+/**
+ * Class used to implement the attachment queue
+ * Requires a PowerSyncDatabase, an implementation of
+ * AbstractRemoteStorageAdapter and an attachment directory name which will
+ * determine which folder attachments are stored into.
+ */
+public open class AttachmentQueue(
```

Comment:
~~Does this have to be `open`? We're not extending it at the moment, and making classes `open` in the future is always easier than the other way around :)~~ Nevermind, saw the logic to create filenames that should be replacable.